tailscale/resource_tailnet_settings.go

```diff
@@ -0,0 +1,140 @@
+	client := m.(*Clients).V2
+
+	settings, err := client.TailnetSettings().Get(ctx)
+	if err != nil {
+		return diagnosticsError(err, "Failed to fetch tailnet settings")
+	}
+
+	settingsMap := map[string]any{
+		"devices_approval_on":                         settings.DevicesApprovalOn,
+		"devices_auto_updates_on":                     settings.DevicesAutoUpdatesOn,
+		"devices_key_duration_days":                   settings.DevicesKeyDurationDays,
+		"users_approval_on":                           settings.UsersApprovalOn,
+		"users_role_allowed_to_join_external_tailnet": string(settings.UsersRoleAllowedToJoinExternalTailnets),
+		"network_flow_logging_on":                     settings.NetworkFlowLoggingOn,
+		"regional_routing_on":                         settings.RegionalRoutingOn,
+		"posture_identity_collection_on":              settings.PostureIdentityCollectionOn,
+	}
+	return setProperties(d, settingsMap)
+}
+
+func resourceTailnetSettingsCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	if err := resourceTailnetSettingsDoUpdate(ctx, d, m); err != nil {
+		return err
+	}
+	d.SetId(createUUID())
+	return resourceTailnetSettingsRead(ctx, d, m)
+}
+
+func resourceTailnetSettingsUpdate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	if err := resourceTailnetSettingsDoUpdate(ctx, d, m); err != nil {
+		return err
+	}
+	return resourceTailnetSettingsRead(ctx, d, m)
+}
+
+func resourceTailnetSettingsDoUpdate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	var role *tsclient.RoleAllowedToJoinExternalTailnets
+	_role, ok := d.GetOk("users_role_allowed_to_join_external_tailnet")
+	if ok {
+		role = tsclient.PointerTo(tsclient.RoleAllowedToJoinExternalTailnets(_role.(string)))
+	}
+	settings := tsclient.UpdateTailnetSettingsRequest{
+		DevicesApprovalOn:                      optional[bool](d, "devices_approval_on"),
+		DevicesAutoUpdatesOn:                   optional[bool](d, "devices_auto_updates_on"),
+		DevicesKeyDurationDays:                 optional[int](d, "devices_key_duration_days"),
+		UsersApprovalOn:                        optional[bool](d, "users_approval_on"),
+		UsersRoleAllowedToJoinExternalTailnets: role,
+		NetworkFlowLoggingOn:                   optional[bool](d, "network_flow_logging_on"),
+		RegionalRoutingOn:                      optional[bool](d, "regional_routing_on"),
+		PostureIdentityCollectionOn:            optional[bool](d, "posture_identity_collection_on"),
+	}
+
+	client := m.(*Clients).V2
+	if err := client.TailnetSettings().Update(ctx, settings); err != nil {
+		return diagnosticsError(err, "Failed to update tailnet settings")
+	}
+
+	return nil
+}
+
+func resourceTailnetSettingsDelete(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	// We don't know what the default values for Tailnet settings should be, so deleting is a noop.
```

Comment:
The "we don't know" confuses me here - technically we do know the defaults we assign to new tailnets. Maybe it's less that we don't know and more that you can't really "delete" settings. (Reverting to defaults could be argued as being "deletion" but also could be argued as not)

Reply:
Maybe I should say "the terraform provider does not know the defaults". There's no API to "get defaults" or "reset to defaults", and I don't think we want to hardcode this in the terraform provider.